Fast dependency scanning for Swift

DougGregor opened the pull request:
Implement a new "fast" dependency scanning option,
`-scan-dependencies`, in the Swift frontend that determines all
of the source file and module dependencies for a given set of
Swift sources. It covers all forms of dependencies:

1) Swift (serialized) module files, by reading the module header
2) Swift interface files, by parsing the source code to find imports
3) Swift source modules, by parsing the source code to find imports
4) Clang modules, using Clang's fast dependency scanning tool
5) Bridging headers, including the source files and modules they depend on

A single `-scan-dependencies` operation maps out the full
dependency graph for the given Swift source files, including all
of the Swift and Clang modules that may need to be built, such
that all of the work can be scheduled up front by the Swift
driver or any other build system that understands this
option. The dependency graph is emitted as JSON, which can be
consumed by these other tools. Swift code with `Codable`
data structures is used for validating this JSON, and can
(eventually) be picked up by other tools.

This operation also supports `-emit-dependencies`, the flag that
produces make-style dependencies to determine all of the input
files that can affect the compilation, making it usable as a fast
dependency scanner even for clients not interesting in prebuilding
modules.

In a source file that imports just `Cocoa` on macOS, fast
dependency scanning with this patch is ~11x faster than
the current dependency scanner (with `-resolve-imports`), taking
it from 5.7s (and writing 48MB of Clang modules) down to .49s.

DougGregor commented:
@Bigcheese this totally doesn't work yet, but it's getting closer to mapping out the various Clang + Swift module builds that are involved in a Swift compile. You can see my super hacky interaction with Clang's fast dependency scanner in `ClangModuleDependencyScanner.cpp`.

DougGregor commented:
Hey, it works a little bit now with a Clang change to handle `@import` directives in Clang's dependency scanner.

DougGregor commented:
The necessary Clang changes are at https://github.com/apple/llvm-project/pull/392

slavapestov commented:
Would this allow the frontend to get out of the business of building interface files altogether?

DougGregor commented:
> Would this allow the frontend to get out of the business of building interface files altogether?

Yes, that's the intent. The driver would coordinate the interface file builds so the frontend wouldn't have to do it.

DougGregor committed:
Fast dependency scanning for Swift

Implement a new "fast" dependency scanning option,
`-scan-dependencies`, in the Swift frontend that determines all
of the source file and module dependencies for a given set of
Swift sources. It covers four forms of modules:

1) Swift (serialized) module files, by reading the module header
2) Swift interface files, by parsing the source code to find imports
3) Swift source modules, by parsing the source code to find imports
4) Clang modules, using Clang's fast dependency scanning tool

A single `-scan-dependencies` operation maps out the full
dependency graph for the given Swift source files, including all
of the Swift and Clang modules that may need to be built, such
that all of the work can be scheduled up front by the Swift
driver or any other build system that understands this
option. The dependency graph is emitted as JSON, which can be
consumed by these other tools.

DougGregor commented:
I collapsed everything because it doesn't make sense for this to ride on top of `-emit-imported-modules`; rather, that flag should (separately) go away. Instead, implement a new flag (`-scan-dependencies`) whose primary output is the JSON dependency graph and which will eventually also emit normal makefile dependencies (`.d` file) as well, for clients that need it.

DougGregor committed:
[Scan dependencies] Find overlays associated with imported Clang modules.

DougGregor committed:
[Dependency scanning] Emit make-style dependency files.

If requested, produce the normal make-style dependency file from the
fast dependency scanner, using the same dependency-tracking
infrastructure.

DougGregor committed:
[Dependency scanning] Scan and record bridging header dependencies.

When there is a bridging header associated with the module, scan and record
its dependencies. Note them in a separate structure to capture the specific
dependencies of the bridging header.

DougGregor committed:
[Dependency scanning] Parse JSON output into Swift data structures.

DougGregor committed:
[Dependency scanning] Add implicit Swift and SwiftOnoneSupport dependencies

DougGregor committed:
[Dependency scanning] Cope with inferred module maps

jerrymarino commented:
So this should this allow us to directly run the frontend in external build systems directly? Seems like we can use the new .d files

DougGregor commented:
> So this should this allow us to directly run the frontend in external build systems directly? Seems like we can use the new .d files

We'd prefer that external build systems go through the driver in general, because the frontend interface is  the supported way to call the compiler. However, we can expose this as a driver option once it's been tested a bit better. I would certainly appreciate if others would test it out and see what I've missed.

futurejones commented:
Can one of the admins verify this patch?

DougGregor commented:
Brought it up to date with master. It no longer needs cherry-picks on top of Clang

DougGregor commented:
@swift-ci please smoke test

DougGregor committed:
Sink the module dependency scanner down into the Serialization library

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please smoke test

DougGregor committed:
[Dependency scanner] Get the Clang module file name from Clang.

DougGregor committed:
[Dependency scanner] Capture Clang context hash and command-line arguments.

davidungar commented:
Lovely, readable code. But, why were the storage classes factored out, instead of just having a base `ModuleDependencies` class and two subclasses? Is there not always a one-to-one relationship to the storage? I'm missing something here...

davidungar commented on lib/ClangImporter/ClangModuleDependencyScanner.cpp:
```diff
@@ -0,0 +1,341 @@
+    return found;
+
+  // Retrieve or create the shared state.
+  auto clangImpl = getOrCreateClangImpl(cache);
```
This function is a bit long; Could this be a `const`?

davidungar commented on lib/ClangImporter/ClangModuleDependencyScanner.cpp:
```diff
@@ -0,0 +1,341 @@
+    return false;
+
+  // Retrieve or create the shared state.
+  auto clangImpl = getOrCreateClangImpl(cache);
```
`const`?

davidungar commented on lib/ClangImporter/ClangModuleDependencyScanner.cpp:
```diff
@@ -0,0 +1,341 @@
+  return cache.findDependencies(moduleName, ModuleDependenciesKind::Clang);
+}
+
+bool ClangImporter::addBridgingHeaderDependencies(
```
You know me; I would prefer to break up this function into more understandable (for me) pieces.

davidungar commented on lib/ClangImporter/ClangModuleDependencyScanner.cpp:
```diff
@@ -0,0 +1,341 @@
+  ClangImporterOptions importerOpts;
+
+  // Retrieve the bridging header.
+  std::string bridgingHeader = *targetModule.getBridgingHeader();
```
`const`?

davidungar commented on lib/ClangImporter/ClangModuleDependencyScanner.cpp:
```diff
@@ -0,0 +1,341 @@
+
+  // Determine the command-line arguments for dependency scanning.
+  auto &ctx = Impl.SwiftContext;
+  std::vector<std::string> commandLineArgs =
```
Sorry if all these  `consts` are getting tedious. As you know, the idea is that if there are more than a few lines after the assignment, if something is `const` the first assignment is enough to know where it's coming from without having to inspect all uses. (Esp. in C++ where one can pass a reference, sigh.) Forgive me for stating the obvious.

davidungar commented on lib/Frontend/Frontend.cpp:
```diff
@@ -729,12 +729,11 @@ void CompilerInstance::addAdditionalInitialImportsTo(
-shouldImplicityImportSwiftOnoneSupportModule(CompilerInvocation &Invocation) {
-  if (Invocation.getImplicitModuleImportKind() !=
-      SourceFile::ImplicitModuleImportKind::Stdlib)
+bool CompilerInvocation::shouldImportSwiftONoneSupport() const {
```
Nice!

davidungar commented on lib/FrontendTool/FrontendTool.cpp:
```diff
@@ -1255,8 +1256,10 @@ static bool performCompile(CompilerInstance &Instance,
     bool CanDelayBodies = Action != FrontendOptions::ActionType::DumpParse;
-    Instance.performParseOnly(/*EvaluateConditionals*/
-                    Action == FrontendOptions::ActionType::EmitImportedModules,
+    bool EvaluateConditionals =
+        Action == FrontendOptions::ActionType::EmitImportedModules
+        || Action == FrontendOptions::ActionType::ScanDependencies;
+    Instance.performParseOnly(EvaluateConditionals,
```
The flow in here and through 1271 is confusing. The predicate here is `FrontendOptions::shouldActionOnlyParse(Action)` and I would expect a return right after line 1262.
But instead, it falls through to 1271: `if (Action == FrontendOptions::ActionType::Parse)` and it that is true it returns. Which leaves me pondering: what's the difference between the two predicates? Is one a subset of the other? There must be a clearer way to write this. Maybe also do a return right here?
OK, I see it, now. Will leave the above confused ramblings in FWIW.

The whole business starting at 1255 could use a comment to the effect of do the the right about of parsing, etc. for the mode, or some such. (I would factor it out into a function with that name.)

Then, a blank line after 1258 would make it clear that the comment above it only applies to that one line.
(Not your fault!)

Then, a "why" comment might be nice about why conditionals are only being evaluated in these two cases.

I guess what's going on is that this function could be split in a "parse" phase, through 1272 and an after-parse phase, and that demarcation would clarify what's going on. (Not your fault, but this change just makes it incrementally worse, like an amoeba that's overdue for fission.)

davidungar reviewed (approved):
I can't for correctness, but overall it looks pretty good. I threw in a few suggestions that could incrementally improve the clarity, IMO. The most important of those would be the places where a "why" comment could help.

DougGregor commented on include/swift/AST/ModuleDependencies.h:
```diff
@@ -0,0 +1,313 @@
+//===----------------------------------------------------------------------===//
+//
+// This file defines data structures for capturing module dependencies.
+//
```
The first part, sure, but this part of the code isn't the "fast" bit---it's just a description of the actual module dependencies.

DougGregor commented on include/swift/AST/ModuleDependencies.h:
```diff
@@ -0,0 +1,313 @@
+  const std::string compiledModulePath;
+
+  /// The set of modules on which this module depends.
+  std::vector<std::string> moduleDependencies;
```
It describes "what has to be built before me". I'll clarify this in the overall file/class comment.

DougGregor commented on lib/AST/ASTContext.cpp:
```diff
@@ -1449,6 +1450,21 @@ void ASTContext::addModuleLoader(std::unique_ptr<ModuleLoader> loader,
+    StringRef moduleName, bool isUnderlyingClangModule,
+    ModuleDependenciesCache &cache) {
+  for (auto &loader : getImpl().ModuleLoaders) {
+    if (isUnderlyingClangModule &&
```
This code stinks; it'll be removed when we refactor more.

DougGregor commented on lib/AST/ModuleDependencies.cpp:
```diff
@@ -0,0 +1,170 @@
+  }
+}
+
+bool ModuleDependenciesCache::hasDependencies(
```
Redundant, IMO.

DougGregor commented on lib/ClangImporter/ClangModuleDependencyScanner.cpp:
```diff
@@ -0,0 +1,341 @@
+  }
+}
+
+llvm::ErrorOr<StringRef> ClangModuleDependenciesCacheImpl::getImportHackFile() {
```
Clang will provide us with an API that makes this unnecessary. It's a HACK.

DougGregor committed:
[Dependency scanning] Improve overview documentation somewhat.

DougGregor committed:
Update test to reflect unimportant change in output results

DougGregor commented:
Rebasing to fix merge conflicts, adding some comments, tweaking tests... and let's get this landed.

DougGregor commented:
@swift-ci please test

DougGregor commented:
@swift-ci please test

DougGregor commented:
@swift-ci please test platform Windows

DougGregor commented:
@swift-ci please test platform Windows

DougGregor commented:
@swift-ci please test
